bastion/src/envelope.rs

```diff
@@ -0,0 +1,131 @@
+use crate::broadcast::Sender;
+use crate::context::BastionContext;
+use crate::message::{BastionMessage, Message, Msg};
+use crate::path::BastionPath;
+use std::sync::Arc;
+
+#[derive(Debug)]
+pub(crate) struct Envelope {
+    pub(crate) msg: BastionMessage,
+    pub(crate) sign: Signature,
+}
+
+/// A tuple containing `Msg` and its `Signature`.
+pub type SignedMessage = (Msg, Signature);
+
+#[derive(Debug, Clone)]
+/// Message signature used to identify message sender and send messages to it.
+///
+/// ```rust
+/// # use bastion::prelude::*;
+/// #
+/// # fn main() {
+///     # Bastion::init();
+///     #
+/// Bastion::children(|children| {
+///     children.with_exec(|ctx: BastionContext| {
+///         async move {
+///             // Wait for a message to be received...
+///             let msg: SignedMessage = ctx.recv().await?;
+///             msg.1.tell(&ctx, "reply").expect("Unable to reply");
+///             Ok(())
+///         }
+///     })
+/// }).expect("Couldn't create the children group.");
+///     #
+///     # Bastion::start();
+///     # Bastion::stop();
+///     # Bastion::block_until_stopped();
+/// # }
+/// ```
+pub struct Signature(SignatureInner);
+
+impl Signature {
+    /// Checks whether the sender is identified.
+    /// Usually anonymous sender means messages sent by
+    /// [broadcast][crate::Bastion::broadcast()] and it's other methods implied to
+    /// be called outside of the Bastion context.
+    pub fn is_anonymous(&self) -> bool {
+        match self.0 {
+            SignatureInner::Anonymous => true,
+            _ => false,
+        }
+    }
+
+    /// Returns `BastionPath` of a sender
+    pub fn sender_path(&self) -> Option<Arc<BastionPath>> {
+        match &self.0 {
+            SignatureInner::Anonymous => None,
+            SignatureInner::Identified(path, _) => Some(path.clone()),
+        }
+    }
+
+    /// Sends message to a message sender on behalf of
+    /// the current `BastionContext`.
+    pub fn tell<M: Message>(&self, ctx: &BastionContext, msg: M) -> Result<(), M> {
```

Comment:
For example, this is an extremely important method: `tell`. Needs an example, because while browsing docs people will skip reading or have problems during their implementation phase.

Reply:
fixed